python-lib/wordcloud_generator.py

```diff
@@ -0,0 +1,193 @@
+        else:
+            # Simply format data similarly
+            self.df_grouped = [(self.language, self.df)]
+        logging.info(f"Data preparation: Done in {(time() - start):.2f} seconds.")
+
+    def _tokenize_texts(self):
+        # Tokenize each group of observations in its correct language
+        start = time()
+        logging.info("Initializing tokenization")
+
+        # Get language and subchart name for each group
+        self.texts = []
+        self.group_names = []
+        for name, group in self.df_grouped:
+            self.texts.append([group[self.text_column].str.cat(sep=" ")])
+            self.group_names.append(name)
+
+        # Get tokenization languages differently depending on language/subchart settings combinations
+        if self.language_column == None and self.subchart_column == None:
+            self.languages = [self.language]
+        elif self.language_column != None and self.subchart_column != None:
+            languages, subcharts = zip(*self.group_names)
+            self.languages = list(languages)
+            self.subcharts = list(subcharts)
+        elif self.subchart_column != None:
+            self.subcharts = self.group_names
+            self.languages = [self.language] * len(self.subcharts)
+        else:
+            self.languages = self.group_names
+
+        # Tokenize
+        self.docs = [
+            self.tokenizer.tokenize_list(text, language)[0] for text, language in zip(self.texts, self.languages)
+        ]
+        logging.info(f"Tokenization done in {(time() - start):.2f} seconds.")
+
+    def _count_tokens(self):
+        # Count tokens in each group
+        start = time()
+        logging.info("Initializing count")
+        self.counters = []
+        for doc in self.docs:
+            counter = Counter()
+            for token in doc:
+                counter[(token.text)] += 1  # Equivalently, token.lemma_
+            self.counters.append(counter)
+        logging.info("Count successful, aggregating counters according to chart settings")
+
+        if self.subchart_column == None:
+            # sum the values with same keys
+            self.counts = Counter()
+            for d in self.counters:
+                self.counts.update(d)
+
+            self.counts = dict(self.counts)
+        else:
+            self.counts_df = pd.DataFrame(list(zip(self.subcharts, self.counters)), columns=["subchart", "count"],)
+            self.counts_df = self.counts_df.groupby(by=["subchart"]).agg({"count": "sum"})
+            # remove subcharts emptied by filter
+            self.counts_df = self.counts_df.loc[self.counts_df["count"] != {}, :]
+
+        logging.info(f"Counter aggregation successful, counting done in {(time() - start):.2f} seconds.")
+
+    def _generate_wordclouds(self):
+        # Generate wordclouds and save them as png images
+        start = time()
+        logging.info("Generating wordclouds")
+        if self.subchart_column != None:
```

Comment:
There is two much similarity between the content of if/else - it should be factorized in a `_generate_wordcloud` method

Reply:
Still some similarity, you could add a small static method to take care of 
```
                # Save chart
                temp = BytesIO()
                fig.savefig(temp, bbox_inches="tight", pad_inches=self.pad_inches, dpi=fig.dpi)
                self.output_folder.upload_data(output_file_name, temp.getvalue())
                plt.close()
```

Reply:
After thinking again about this, should be replaced by a `yield (output_file_name, bytes_buffer)` which is then consumed and sent to the folder in the recipe.

That way no need for the dataiku API in this class

Reply:
After checking, I am not sure this is specific enough to warrant a `WordcloudExporter` class. The class is mostly empty. Since this is a fairly generic and simple piece of code, I would rather have just one `WordcloudGenerator` class which yields pairs of path, bytes. And in `recipe.py` consume that generator to write to the output folder.